netket/optimizer/jax_stochastic_reconfiguration.py

```diff
@@ -0,0 +1,370 @@
+        sparse_maxiter=None,
+    ):
+
+        self._lsq_solver = lsq_solver
+        self._diag_shift = diag_shift
+        self._use_iterative = use_iterative
+        self._is_holomorphic = is_holomorphic
+        self._svd_threshold = svd_threshold
+        self._scale_invariant_pc = False
+        self._S = None
+        self._last_rank = None
+
+        # Temporary arrays
+        self._v_tilde = None
+        self._res_t = None
+
+        # Quantities for sparse solver
+        self.sparse_tol = 1.0e-5 if sparse_tol is None else sparse_tol
+        self.sparse_maxiter = sparse_maxiter
+        self._lsq_solver = lsq_solver
+        self._x0 = None
+        self._init_solver()
+
+        self._comm = MPI.COMM_WORLD
+
+    def _init_solver(self):
+        lsq_solver = self._lsq_solver
+
+        if lsq_solver in ["gmres", "cg", "minres","jaxcg"]:
+            self._use_iterative = True
+        if lsq_solver in ["ColPivHouseholder", "QR", "SVD", "Cholesky"]:
+            self._use_iterative = False
+
+        if self._use_iterative:
+            if lsq_solver is None:
+                lsq_solver = "jaxcg"
+
+            if lsq_solver == "gmres":
+                self._sparse_solver = partial(gmres, atol="legacy")
+            elif lsq_solver == "cg":
+                self._sparse_solver = partial(cg, atol="legacy")
+            elif lsq_solver == "minres":
+                if self._is_holomorphic:
+                    raise RuntimeError(
+                        "minres can be used only for real-valued parameters."
+                    )
+                self._sparse_solver = minres
+            elif not lsq_solver == 'jaxcg':
+                raise RuntimeError("Unknown sparse lsq_solver " + lsq_solver + ".")
+
+        else:
+            if (
+                lsq_solver is None
+                or "ColPivHouseholder" in lsq_solver
+                or "QR" in lsq_solver
+            ):
+                self._lapack_driver = "gelsy"
+            elif "SVD" in lsq_solver or "Cholesky" in lsq_solver:
+                self._lapack_driver = None
+            else:
+                self._lapack_driver = None
+                raise RuntimeError("Unknown lsq_solver" + lsq_solver + ".")
+
+        if self._use_iterative and self._svd_threshold is not None:
+            raise ValueError(
+                "The svd_threshold option is available only for non-sparse solvers."
+            )
+
+    def compute_update(self, oks, grad, out=None):
+        r"""
+        Solves the SR flow equation for the parameter update ẋ.
+
+        The SR update is computed by solving the linear equation
+           Sẋ = f
+        where S is the covariance matrix of the partial derivatives
+        O_i(v_j) = ∂/∂x_i log Ψ(v_j) and f is a generalized force (the loss
+        gradient).
+
+        Args:
+            oks: The matrix of log-derivatives,
+		O_i(v_j)
+            grad: The vector of forces f.
+            out: Output array for the update ẋ.
+        """
+
```

Comment:
basicaly, you can put here the first call you have now in the vmc that prepares the oks and the grad to the right format

Reply:
Right, is it okay to call the raw jacobian oks as well? I don't want to use two names to avoid copying memory.

Reply:
yes it's fine